### PR TITLE
gh-95415: Make availability directive consistent (GH-95416)

### PR DESCRIPTION
**Pull request by tiran**
* Issue: gh-95415

### CODE REVIEW

File: Doc/library/os.rst [hunk 1]
```diff
@@ -824,7 +824,7 @@ as internal buffering of data.
       It will always copy no bytes and return 0 as if the file was empty
       because of a known Linux kernel issue.
 
-   .. availability:: Linux kernel >= 4.5 or glibc >= 2.27.
+   .. availability:: Linux >= 4.5, glibc >= 2.27.
```
Comment (encukou) [inline Doc/library/os.rst]: Does this mean “Linux >= 2.6.17 **or** glibc >= 2.5”, or “Linux >= 2.6.17 **with** glibc >= 2.5”?
From the `memfd_create` note it looks like it's the latter, so the comma (which means “or” everywhere else) would be wrong.

File: Doc/library/os.rst [hunk 2]
```diff
@@ -1164,7 +1164,7 @@ or `the MSDN <https://msdn.microsoft.com/en-us/library/z0kc8e3z.aspx>`_ on Windo
    descriptors are :ref:`non-inheritable <fd_inheritance>`. For a (slightly) more
    portable approach, use the :mod:`pty` module.
 
-   .. availability:: some flavors of Unix.
+   .. availability:: Unix.
```
Comment (encukou) [inline Doc/library/os.rst:1167]: This is losing information. I don't know if it's deliberate here.
Would it be possible to use `Unix (some flavors)`?
Or `Unix*`, meaning “see note below” for humans and “don't fail the test” for future tests?

File: Doc/tools/extensions/pyspecific.py [hunk 3]
```diff
@@ -147,8 +157,48 @@ def run(self):
...
+            elif not version:
+                version = True
+            platforms[platform] = version
+
+        unknown = set(platforms).difference(self.known_platforms)
+        if unknown:
+            logger = logging.getLogger(type(self).__qualname__)
+            logger.warn(
+                f"Unknown platform(s) '{' '.join(sorted(unknown))}' in "
+                f"'.. availability:: {self.arguments[0]}'."
```
Comment (encukou) [inline Doc/tools/extensions/pyspecific.py]: Please add a comment about what to do if this test faills. Devs can just add a platform to the list if it's missing, right?

**Review by encukou — COMMENTED** [review]
Big +1 on the idea, some nitpicks:

File: Doc/library/os.rst (hunk 1 above)
Comment (tiran) [reply Doc/library/os.rst]: Python needs both the Linux Kernel syscall and the glibc wrapper to implement the feature. So both conditions must apply.

There are multiple options to handle this:
- We could either clarify this in https://docs.python.org/3/library/intro.html#availability and explain that for a combination of OS and libc, both conditions must hold
- or we could extend the code of the directive class
- or add a remark to the body of the availability directive.

File: Doc/library/os.rst (hunk 2 above)
Comment (tiran) [reply Doc/library/os.rst:1167]: It is deliberate. Our documentation already states that “Availability: Unix” means "some Unix but not all Unix-like platforms":

> An “Availability: Unix” note means that this function is commonly found on Unix systems.

File: Doc/tools/extensions/pyspecific.py (hunk 3 above)
Comment (tiran) [reply Doc/tools/extensions/pyspecific.py]: Done

File: Doc/library/os.rst (hunk 1 above)
Comment (encukou) [reply Doc/library/os.rst]: What about accepting ` with `, which would just be an alias for `,` for this check?

Comment (tiran) [reply Doc/library/os.rst]: The code does not use the return value of `parse_platforms`. Let's ignore the `with` for now. `Linux >= 3.17 with glibc >= 2.27` is parsed as `{"Linux": "3.17 with glibc >= 2.27"}` and the with part is not validated.

**Review by encukou — APPROVED** [review]
Looks good!

File: Doc/library/socket.rst [hunk 4]
```diff
@@ -165,15 +165,19 @@ created.  Socket addresses are represented as follows:
...
+     Some algorithm types require more recent Kernels.
 
   .. versionadded:: 3.6
 
 - :const:`AF_VSOCK` allows communication between virtual machines and
   their hosts. The sockets are represented as a ``(CID, port)`` tuple
   where the context ID or CID and port are integers.
 
-  .. availability:: Linux >= 4.8 QEMU >= 2.8 ESX >= 4.0 ESX Workstation >= 6.5.
+  .. availability:: Linux >= 3.9
```
Comment (ezio-melotti) [inline Doc/library/socket.rst:178]: Is this correct?

File: Doc/library/socket.rst [hunk 5]
```diff
@@ -1147,7 +1153,9 @@ The :mod:`socket` module also offers various network-related services:
    amount of ancillary data that can be received, since additional
    data may be able to fit into the padding area.
 
-   .. availability:: most Unix platforms, possibly others.
+   .. availability:: Unix.
+
+      most Unix platforms.
```
Comment (ezio-melotti) [inline Doc/library/socket.rst:1158]:
```suggestion
      Most Unix platforms.
```

File: Doc/library/threading.rst [hunk 6]
```diff
@@ -429,7 +431,7 @@ since it is impossible to detect the termination of alien threads.
          system-wide) from the time the thread is created until the thread
          has been terminated.
 
-      .. availability:: Requires :func:`get_native_id` function.
+      .. availability:: Windows, FreeBSD, Linux, macOS, OpenBSD, NetBSD, AIX, DragonFlyBSD.
```
Comment (ezio-melotti) [inline Doc/library/threading.rst:434]: I would keep the previous sentence under the list of platforms.

File: Doc/tools/extensions/pyspecific.py [hunk 7]
```diff
@@ -134,11 +134,21 @@ def run(self):
...
 
+    known_platforms = frozenset({
+        "AIX", "Android", "BSD", "DragonFlyBSD", "Emscripten", "FreeBSD",
+        "Linux", "NetBSD", "OpenBSD", "POSIX", "Solaris", "Unix", "VxWorks",
+         "WASI", "Windows", "macOS",
+         # libc
+         "BSD libc", "glibc", "musl",
+         # POSIX platforms with pthreads
+         "pthreads",
+    })
```
Comment (ezio-melotti) [inline Doc/tools/extensions/pyspecific.py:151]:
```suggestion
    known_platforms = frozenset({
        "AIX", "Android", "BSD", "DragonFlyBSD", "Emscripten", "FreeBSD",
        "Linux", "NetBSD", "OpenBSD", "POSIX", "Solaris", "Unix", "VxWorks",
        "WASI", "Windows", "macOS",
        # libc
        "BSD libc", "glibc", "musl",
        # POSIX platforms with pthreads
        "pthreads",
    })
```

File: Doc/library/os.rst [hunk 8]
```diff
@@ -706,7 +706,7 @@ process and user.
    :func:`socket.gethostname`  or even
    ``socket.gethostbyaddr(socket.gethostname())``.
 
-   .. availability:: recent flavors of Unix.
+   .. availability:: Unix.
```
Comment (ezio-melotti) [inline Doc/library/os.rst:709]: Should these get a note like `Might not be available on less recent flavors of Unix.` or `Might not be available on some flavors of Unix.`?  Even better would be to be more specific (e.g. `Not available on Unix flavors that ...`).

**Review by ezio-melotti — COMMENTED** [review]
A bit late to the party since the PR already got merged, but here is my review.

File: Doc/library/socket.rst (hunk 4 above)
Comment (tiran) [reply Doc/library/socket.rst:178]: yes, the previous information was incorrect. vsock was first provided in Linux 3.9, https://man7.org/linux/man-pages/man7/vsock.7.html

File: Doc/library/os.rst (hunk 8 above)
Comment (tiran) [reply Doc/library/os.rst:709]: `uname` is available on all Unix-like platforms that we support. It was standardized in IEEE Std 1003.1-2001, https://pubs.opengroup.org/onlinepubs/009604599/functions/uname.html . Every POSIX-like OS released in the last 20 years should have it.

**Pull request merged**